### PR TITLE
Close dialogs on history back

### PR DESCRIPTION
**Pull request by bramkragten**
## Proposed change

A start on closing dialogs on history back like we already did for the more info dialog.

Extra bonus, going forward in the browser history will reopen the dialog (if possible).
If you open a dialog from another dialog, going back from the second dialog will reopen the first dialog.

Dialogs should implement a `closeDialog` function, that should take care of closing the dialog and then fire the event `dialog-closed`.

You can prevent a dialog from closing on back by returning false from `closeDialog` (for a dialog that should not be closed without a specific action of that doesn't have a cancel/close button).

Dialogs that don't implement `closeDialog` yet, won't be closed on back, and the history will be a little screwed up (first back will not do anything, second back will give the current behavior).
So we should add `closeDialog` to all dialogs 😄 

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/dialogs/config-flow/dialog-data-entry-flow.ts [hunk 1]
```diff
@@ -295,13 +305,10 @@ class DataEntryFlowDialog extends LitElement {
...
-    if (this._step) {
-      this._flowDone();
-    } else if (this._step === null) {
-      // Flow aborted during picking flow
-      this._step = undefined;
-      this._params = undefined;
+  private _close(ev: CustomEvent) {
+    // If closed is called because of disconnect of ha-dialog, we don't have to handle
+    if ((ev.target as HaDialog).isConnected) {
+      history.back();
```
Comment (balloob) [inline src/dialogs/config-flow/dialog-data-entry-flow.ts:311]: `history.back` seems weird here. The dialog should not be aware on how these things work. Maybe we can add an event to dialogs to indicate when they are closing?

File: src/dialogs/more-info/ha-more-info-dialog.ts [hunk 2]
```diff
@@ -231,11 +234,14 @@ export class MoreInfoDialog extends LitElement {
...
-    this._close();
+    this.closeDialog();
   }
 
-  private _close() {
-    fireEvent(this, "hass-more-info", { entityId: null });
+  private _close(ev: CustomEvent) {
+    // If closed is called because of disconnect of ha-dialog, we don't have to handle
+    if ((ev.target as HaDialog).isConnected) {
+      history.back();
```
Comment (balloob) [inline src/dialogs/more-info/ha-more-info-dialog.ts:243]: Same here.

File: src/state/url-sync-mixin.ts [hunk 3]
```diff
@@ -12,81 +17,76 @@ export const urlSyncMixin = <T extends Constructor<HassBaseEl>>(
...
-
-          if (this._ignoreNextHassChange) {
-            if (DEBUG) {
-              console.log("ignore hasschange");
-            }
-            this._ignoreNextHassChange = false;
-            return;
-          }
+        private _dialogClosedListener = (ev: HASSDomEvent<undefined>) => {
+          const dialogElement = ev.composedPath()[0] as Element;
```
Comment (balloob) [inline src/state/url-sync-mixin.ts:35]: Would it be cleaner if we added this to the event details?

**Review by balloob — APPROVED** [review]

File: src/state/url-sync-mixin.ts (hunk 3 above)
Comment (bramkragten) [reply src/state/url-sync-mixin.ts:35]: I had that before, but is very error prone and a lot more work to implement it all 🤷‍♂️

Comment (balloob) [reply src/state/url-sync-mixin.ts:35]: How can it be error prone? We fire it from the dialog element, so could use this.localName when passing it to `fireEvent` ?

Comment (bramkragten) [reply src/state/url-sync-mixin.ts:35]: true...

**Pull request merged**